hpx/components/process/util/posix/executor.hpp

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2006, 2007 Julio M. Merino Vidal
+// Copyright (c) 2008 Ilya Sokolov, Boris Schaeling
+// Copyright (c) 2009 Boris Schaeling
+// Copyright (c) 2010 Felipe Tanus, Boris Schaeling
+// Copyright (c) 2011, 2012 Jeff Flinn, Boris Schaeling
+// Copyright (c) 2016 Hartmut Kaiser
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_PROCESS_POSIX_EXECUTOR_HPP
+#define HPX_PROCESS_POSIX_EXECUTOR_HPP
+
+#include <hpx/config.hpp>
+
+#if !defined(HPX_WINDOWS)
+#include <hpx/util/unused.hpp>
+#include <hpx/components/process/util/child.hpp>
+
+#include <cstdlib>
+
+#include <sys/types.h>
+#include <unistd.h>
+
+namespace hpx { namespace components { namespace process { namespace posix {
+
+struct executor
+{
+    executor() : exe(0), cmd_line(0), env(0) {}
+
+    struct call_on_fork_setup
+    {
+        executor &e_;
+
+        call_on_fork_setup(executor &e) : e_(e) {}
+
+        template <class Arg>
+        void operator()(const Arg &arg) const
+        {
+            arg.on_fork_setup(e_);
+        }
+    };
+
+    struct call_on_fork_error
+    {
+        executor &e_;
+
+        call_on_fork_error(executor &e) : e_(e) {}
+
+        template <class Arg>
+        void operator()(Arg &arg) const
+        {
+            arg.on_fork_error(e_);
+        }
+    };
+
+    struct call_on_fork_success
+    {
+        executor &e_;
+
+        call_on_fork_success(executor &e) : e_(e) {}
+
+        template <class Arg>
+        void operator()(Arg &arg) const
+        {
+            arg.on_fork_success(e_);
+        }
+    };
+
+    struct call_on_exec_setup
+    {
+        executor &e_;
+
+        call_on_exec_setup(executor &e) : e_(e) {}
+
+        template <class Arg>
+        void operator()(Arg &arg) const
+        {
+            arg.on_exec_setup(e_);
+        }
+    };
+
+    struct call_on_exec_error
+    {
+        executor &e_;
+
+        call_on_exec_error(executor &e) : e_(e) {}
+
+        template <class Arg>
+        void operator()(Arg &arg) const
+        {
+            arg.on_exec_error(e_);
+        }
+    };
```

Comment:
I think all of the above parameters should be changed to take arg by const reference as operator() takes them by const reference as well.